csrc/multidevice/utils.cpp

```diff
@@ -13,22 +13,93 @@
 #include <multidevice/lower_communication.h>
 #include <multidevice/utils.h>
 #include <ops/all_ops.h>
+#include <root_domain_map.h>
 #include <scheduler/utils.h>
 
 #include <c10/util/irange.h>
 
 namespace nvfuser {
 namespace {
 
-std::vector<IterDomain*> getShardedIterDomains(TensorView* tv) {
-  std::vector<IterDomain*> sharded_ids;
+std::unordered_set<IterDomain*> getShardedIterDomains(TensorView* tv) {
+  std::unordered_set<IterDomain*> sharded_ids;
   std::copy_if(
       tv->getLeafDomain().begin(),
       tv->getLeafDomain().end(),
-      std::back_inserter(sharded_ids),
+      std::inserter(sharded_ids, sharded_ids.begin()),
       [](auto id) { return id->isDeviceDim(); });
   return sharded_ids;
 }
+
+// Returns whether a IterDomain in a TensorView can be read/written
+// contiguously. when a IterDomain's sharding is flipped. This occurs when the
+// IterDomain being unsharded/sharded is the outermost in the allocation domain.
+// e.g. [DIDx(i0), r1, i2, i3]
+// i0 => true, outermost axis
+// i1 => true, reduction axis are not allocated
+// i2 => true, reduction and device axis are not allocated
+// i3 => false, i2 is allocated.
+bool isContiguousShard(TensorView* tv, IterDomain* changed_id) {
```

Comment:
This seems a little confusing to me. IIUC, looks like what you're answering here is whether each per-device tensor corresponds to a logically contiguous part of the overall tensor. If that's the case, no packing/permutation is necessary, so you can skip the insertion of the permutation below.

Does my understanding make sense? If so, this is a property of a tensor, not an IterDomain of the tensor, so having a parameter like `changed_id` doesn't seem necessary.

Reply:
> if sharded extent of i0 == 1.

Isn't it always the case?

Reply:
Sorry, my example wasn't very clear. I was thinking when we support sharding on a split axis we can have a scenario like this:

For a 2D tensor sharded on DeviceMesh [Dx, Dy] where axis0 is ahrded on Dx, axis1 sharded on Dy.

root [i0, i1] 
leaf [DIDx(Dx), i0/Dx, DIDy(Dy), i1/Dy] 
allocation [i0/Dx, i1/Dy]

if axis 1 is unsharded:
root [i0, i1] 
leaf [DIDx(Dx), i0/Dx, Dy, i1/Dy]
allocation [i0/Dx, i1]

To check if we need a permutation for this pattern, we could either check leaf domain index 3 is the outermost non-device/non-reduce axis (false). 
Or check from that index 1 on the allocation domain is the outermost (false). This check is a bit cleaner.

Reply:
Thanks. That makes sense.

Reply:
Is `changed_id` always parallelized with DID?

Reply:
No, I assume it is either serial or DID.

Reply:
To clarify why the parallel type of `changed_id` isn't important, we just want to know if `changed_id` is the outermost allocated axis. 
Sorry, the comment and function name are very bad. I should call this function `isOutermostAllocatedId(TensorView* tv, IterDomain* id)`